### ISSUE TITLE
Einsum as result of autograd.Function forward makes that backward is not called

### ISSUE THREAD

**Issue by pimdh** [pytorch/pytorch#22072]
## 🐛 Bug

This is a more general version of the bug in #21970, which I'll close.

When a `autograd.Function` returns the result of `einsum`, the backward pass is ignored.
When the result of `einsum` is `clone`d before returning, the backward pass works fine.

## To Reproduce
```
import torch
from torch.autograd import Function


class TestFunction(Function):
    @staticmethod
    def forward(ctx, x, w):
        ctx.save_for_backward(x, w)
        y = torch.einsum('ab,ab->a', [x, w])
        return y

    @staticmethod
    def backward(ctx, grad):
        x, w = ctx.saved_tensors
        w_grad = torch.einsum('a,ab->ab', [grad, x])
        x_grad = torch.einsum('a,ab->ab', [grad, w])
        return x_grad, w_grad


class TestFunction2(Function):
    @staticmethod
    def forward(ctx, x, w):
        ctx.save_for_backward(x, w)
        y = torch.einsum('ab,ab->a', [x, w])
        return y.clone()

    @staticmethod
    def backward(ctx, grad):
        x, w = ctx.saved_tensors
        w_grad = torch.einsum('a,ab->ab', [grad, x])
        x_grad = torch.einsum('a,ab->ab', [grad, w])
        return x_grad, w_grad


x = torch.ones(1, 1)
w = torch.ones(1, 1).requires_grad_()
y = TestFunction.apply(x, w)
z = y.sum()
z.backward()
print(y, z, w.grad)

x = torch.ones(1, 1)
w = torch.ones(1, 1).requires_grad_()
y = TestFunction2.apply(x, w)
z = y.sum()
z.backward()
print(y, z, w.grad)
```

## Expected behavior
The first output should equal the second output, but instead, we get:
```
tensor([1.], grad_fn=<AsStridedBackward>) tensor(1., grad_fn=<SumBackward0>) None
tensor([1.], grad_fn=<TestFunction2Backward>) tensor(1., grad_fn=<SumBackward0>) tensor([[1.]])
```

## Environment
```
PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 2080 Ti
Nvidia driver version: 410.104
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.16.3
[pip] pytorch-ignite==0.2.0
[pip] pytorch-memlab==0.0.3
[pip] torch==1.1.0
[pip] torchvision==0.2.2
[conda] blas                      1.0                         mkl  
[conda] ignite                    0.2.0                    py37_0    pytorch
[conda] mkl                       2019.3                      199  
[conda] mkl_fft                   1.0.12           py37ha843d7b_0  
[conda] mkl_random                1.0.2            py37hd81dba3_0  
[conda] pytorch                   1.1.0           py3.7_cuda10.0.130_cudnn7.5.1_0    pytorch
[conda] pytorch-memlab            0.0.3                    pypi_0    pypi
[conda] torchvision               0.2.2                      py_3    pytorch
```
## Additional context

When I attach a breakpoint at `TestFuction.backward` this is never triggered, but a breakpoint at `TestFunction2.backward` is.

**Comment by t-vi** [comment]
Thank you for the report!
That's bad, note that cloning is not necessary, but even the (no-op) `return y.view(*y.shape)` works. Apparently there is some flagging not done when using as_strided, I'll have to check why that happens.

**Comment by t-vi** [comment]
So probably the best workaround is to do `ctx.mark_dirty(y)` in the forward.
I'll fix it by to replacing the `squeeze_`s at the end of `einsum` and `sumproduct_pair` with a `.view`.

**Comment by pimdh** [comment]
Thanks for looking into it!
Interestingly, `ctx.mark_dirty(y)` does solve the issue in the above example, but not in my actual experiments. `y = y.view(*y.shape)` does solve the issue. I'll see if I can reproduce it in a simple example that I can post here.

**Comment by t-vi** [comment]
If `y.detach()` works, it might be cleaner.
Discussing the PR #22111 with Adam, he thinks the autograd should handle this more gracefully (rather than the grad itself being the problem).

**Comment by pimdh** [comment]
Yes, `y.detach()` works. Thanks.
